File: 05-go-sse/x.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"time"
)

// StartX polls the X/Twitter recent-search API every ~12s and streams new
// tweets as the X feed. If no bearer token is set it does nothing (no crash).
func StartX(ctx context.Context, hub *Hub, bearer, mode, target string) {
	if bearer == "" || target == "" {
		log.Printf("[x] disabled (need X_BEARER_TOKEN + X_TARGET)")
		return
	}

	var query string
	switch mode {
	case "replies":
		query = "conversation_id:" + target
	case "mentions":
		query = "@" + target + " -is:retweet"
	case "hashtag":
		query = "#" + target + " -is:retweet"
	default:
		query = "@" + target + " -is:retweet"
		mode = "mentions"
	}
	log.Printf("[x] polling mode=%s target=%s", mode, target)

	go func() {
		var sinceID string
		backoff := 12 * time.Second
		ticker := time.NewTimer(0)
		defer ticker.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
			}

			newSince, rateLimited, err := pollX(ctx, hub, bearer, query, sinceID)
			if err != nil {
				log.Printf("[x] poll error: %v", err)
			}
			if newSince != "" {
				sinceID = newSince
			}

			next := 12 * time.Second
			if rateLimited {
				backoff *= 2
				if backoff > 5*time.Minute {
					backoff = 5 * time.Minute
				}
				next = backoff
				log.Printf("[x] rate limited; backing off %s", next)
			} else {
				backoff = 12 * time.Second
			}
			ticker.Reset(next)
		}
	}()
}

func pollX(ctx context.Context, hub *Hub, bearer, query, sinceID string) (string, bool, error) {
	q := url.Values{}
	q.Set("query", query)
	q.Set("max_results", "100")
	q.Set("tweet.fields", "created_at,author_id")
	q.Set("expansions", "author_id")
	q.Set("user.fields", "username")
	if sinceID != "" {
		q.Set("since_id", sinceID)
	}
	endpoint := "https://api.twitter.com/2/tweets/search/recent?" + q.Encode()

	reqCtx, cancel := context.WithTimeout(ctx, 15*time.Second)
	defer cancel()
	req, _ := http.NewRequestWithContext(reqCtx, http.MethodGet, endpoint, nil)
	req.Header.Set("Authorization", "Bearer "+bearer)

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", false, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)

	if resp.StatusCode == 429 {
		return "", true, fmt.Errorf("429 rate limited")
	}
	if resp.StatusCode != 200 {
		return "", false, fmt.Errorf("status %d: %s", resp.StatusCode, truncate(string(body), 200))
	}

	var parsed struct {
		Data []struct {
			ID       string `json:"id"`
			Text     string `json:"text"`
			AuthorID string `json:"author_id"`
		} `json:"data"`
		Includes struct {
			Users []struct {
				ID       string `json:"id"`
				Username string `json:"username"`
			} `json:"users"`
		} `json:"includes"`
		Meta struct {
			NewestID string `json:"newest_id"`
		} `json:"meta"`
	}
	if err := json.Unmarshal(body, &parsed); err != nil {
		return "", false, err
	}

	users := map[string]string{}
	for _, u := range parsed.Includes.Users {
		users[u.ID] = u.Username
	}

	// data[] is newest-first -> reverse to chronological.
	for i := len(parsed.Data) - 1; i >= 0; i-- {
		t := parsed.Data[i]
		uname := users[t.AuthorID]
		if uname == "" {
			uname = "user"
		}
		hub.Publish(Message{
			ID:      "x-" + t.ID,
			Source:  "x",
			Channel: query,
			Author:  uname,
			Text:    stripNewlines(t.Text),
			Color:   "",
			TS:      nowMs(),
		})
	}

	return parsed.Meta.NewestID, false, nil
}

func truncate(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}

File: 02-websocket-hub/server/connectors/x.js
'use strict';

const { makeMessage } = require('../schema');

const SEARCH_URL = 'https://api.twitter.com/2/tweets/search/recent';
const POLL_MS = 12000;

// X / Twitter polling connector. No live chat — we poll recent search and stream new tweets.
// mode = replies | mentions | hashtag ; target = conversation id / handle / tag.
function startX({ bearer, mode, target }, emit, log) {
  if (!bearer) {
    log('x: no X_BEARER_TOKEN — connector skipped.');
    return () => {};
  }
  if (!target) {
    log('x: no X_TARGET configured — connector skipped.');
    return () => {};
  }

  let closed = false;
  let sinceId = null;
  let timer = null;

  function buildQuery() {
    if (mode === 'mentions') return `@${target} -is:retweet`;
    if (mode === 'hashtag') return `#${target} -is:retweet`;
    return `conversation_id:${target}`; // default: replies
  }

  async function poll() {
    if (closed) return;
    const params = new URLSearchParams({
      query: buildQuery(),
      max_results: '100',
      'tweet.fields': 'created_at,author_id',
      expansions: 'author_id',
      'user.fields': 'username',
    });
    if (sinceId) params.set('since_id', sinceId);

    let nextDelay = POLL_MS;
    try {
      const res = await fetch(`${SEARCH_URL}?${params.toString()}`, {
        headers: { Authorization: `Bearer ${bearer}` },
      });

      if (res.status === 429) {
        nextDelay = 60000; // rate limited — back off, keep running
        log('x: 429 rate-limited, backing off 60s');
      } else if (!res.ok) {
        nextDelay = 30000;
        log(`x: search returned ${res.status}, retrying in 30s`);
      } else {
        const json = await res.json();
        const tweets = (json && json.data) || [];
        const users = {};
        const inc = json && json.includes && json.includes.users;
        if (inc) for (const u of inc) users[u.id] = u.username;

        // data[] is newest-first -> reverse to chronological
        const chrono = tweets.slice().reverse();
        for (const t of chrono) {
          if (!sinceId || BigInt(t.id) > BigInt(sinceId)) sinceId = t.id;
          emit(
            makeMessage({
              id: `x-${t.id}`,
              source: 'x',
              channel: `${mode || 'replies'}:${target}`,
              author: users[t.author_id] || t.author_id || 'unknown',
              text: t.text,
              color: '',
              ts: t.created_at ? Date.parse(t.created_at) || Date.now() : Date.now(),
            })
          );
        }
        if (chrono.length) log(`x: +${chrono.length} new tweet(s) (since_id=${sinceId})`);
      }
    } catch (err) {
      nextDelay = 30000;
      log(`x: poll error ${err.message}, retrying in 30s`);
    }

    if (!closed) timer = setTimeout(poll, nextDelay);
  }

  log(`x: polling every ${POLL_MS / 1000}s — mode=${mode || 'replies'} target=${target}`);
  poll();

  return function stop() {
    closed = true;
    if (timer) clearTimeout(timer);
  };
}

module.exports = { startX };
